addons/portal/controllers/thread.py

```diff
@@ -0,0 +1,21 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request
+from odoo.addons.mail.controllers import thread
+from odoo.addons.portal.utils import validate_thread_with_hash_pid, validate_thread_with_token
+
+
+class ThreadController(thread.ThreadController):
+
+    def _prepare_post_data(self, post_data, thread, **kwargs):
+        post_data = super()._prepare_post_data(post_data, thread, **kwargs)
+        if request.env.user._is_public():
+            if validate_thread_with_hash_pid(thread, kwargs.get("hash"), kwargs.get("pid")):
+                post_data["author_id"] = int(kwargs["pid"])
+            elif (
+                (partner_fields := thread._mail_get_partner_fields())
+                and thread._fields[partner_fields[0]].type == "many2one"
```

Comment:
I don't really know the feature, but shouldn't you take the first field that is a many2one instead?
Or allow x2many too? Possibly taking the first record of the first non-empty x2many?

@tde-banana-odoo how should the order of fields and x2many fields be considered here?

Reply:
If you want the partner you can use ``_mail_get_partners``, and take the first one. In standard code it is not yet used everywhere, main use cases are employee, pos order, partner (as it returns himself) and studio partner. Idea is to consider that we use the first one is somehow you return a 2many, which is somehow not the standard flow (but should not crash, that way we let people use the mail framework as they want).